source/generators/MuonAngleGenerator.cc

```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
 
 void MuonAngleGenerator::GetDirection(G4ThreeVector& dir)
 {
-  // GetAngles from file?? Azimuth defined anticlockwise
-  // From north
-  G4double zenith  = 0.;
-  G4double azimuth = 0.;
-  distribution_->GetRandom2(azimuth, zenith);
-  // !! Current distribution in units of pi
-  zenith  *= pi;
-  azimuth *= pi;
-
-  dir.setX(sin(zenith) * sin(azimuth));
-  dir.setY(-cos(zenith));
-  dir.setZ(-sin(zenith) * cos(azimuth));
-
-  dir *= *rPhi_;
+
+  // Bool to check if zenith has a valid value. If not then resample
+  G4bool invalid_evt = true;
+
+  while(invalid_evt){
+  
+    // Amount to smear the randomly sampled zenith/azimuth values by
+    G4double zen_BW_smear{1.0e6}; 
+    G4double az_BW_smear{1.0e6};
+
+    // Generate random index weighted by the bin contents
+    G4int RN_indx = discr_dist_(RN_engine_);
+
+    // Loop over the zenith values and find the corresponding bin width to smear
+    for (G4int i = 0; i < zenith_bins_.size()-1; i++){
+        
+        // Catch very last bin
+        if (zen_BW_smear == 1e6 && i == zenith_bins_.size()-2){
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+              && zeniths_[RN_indx] <= zenith_bins_[i+1]){
+            
+              zen_BW_smear = zen_BW_[i];
+            }
+        }
+        else {
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+                && zeniths_[RN_indx] < zenith_bins_[i+1]){
+              
+              zen_BW_smear = zen_BW_[i];
+
+          }
+        }
+    
+    }
+
+    // Loop over the azimuth values and find the corresponding bin width to smear
+    for (G4int i = 0; i < azimuth_bins_.size()-1; i++){
+        
+        // Include last bin in check
+        if (az_BW_smear == 1e6 && i == azimuth_bins_.size()-2){
+          if (azimuths_[RN_indx] >= azimuth_bins_[i] 
+              && azimuths_[RN_indx] <= azimuth_bins_[i+1]){
+            
+              az_BW_smear = az_BW_[i];
+
+          }
+        }
+        else {
+
+          if (azimuths_[RN_indx] >= azimuth_bins_[i] 
+                && azimuths_[RN_indx] < azimuth_bins_[i+1]){
+              
+              az_BW_smear = az_BW_[i];
+
+          }
+        }
+
+    }
+
+    // Check if the smear values are set properly
+    if (az_BW_smear == 1.0e6 || zen_BW_smear == 1.0e6 )
+      G4Exception("[MuonAngleGenerator]", "GetDirection()",
+                FatalException, " bin smear value was not set correctly ");
```

Comment:
Following previous comment, if smear values are -1, it would mean that the sampled value (`azimuths_[RN_indx], zeniths_[RN_indx]`) is not inside any of the bins in the input file. Therefore I think a message such that "Sampled angle not inside binning, review input file" will be helpfull.

Reply:
We might also avoid doing this check for every sampling, and add a check above such that
`azimuth_bins_[0] <= azimuths_.min() && azimuths_.max() <=  azimuth_bins_[-1]`. This also avoids to define meaninless `az_BW_smear`. Same for zenith

Reply:
Updated error message to be more descriptive. 

I do not quite understand what you mean by the second check here, happy to follow up with you on slack about this.

Reply:
Defining a meaningless `smearVal` is only used to check is the sampled `sampVal` is inside any of the bins provided at the configuration file, explicitly:

```
    // Check if the smear values are set properly
    if (az_BW_smear == std::numeric_limits<G4double>::lowest() ||
        zen_BW_smear == std::numeric_limits<G4double>::lowest() ){
                G4Exception("[MuonAngleGenerator]", "GetDirection()",
                FatalException, " Sampled angle not inside binning, review input file ");
    }
```

My point is that instead doing this check at each call to `GetDirection`, doing it globally in `LoadMuonDistribution` with something like (pseudocode):
`azimuth_bins_[0] <= min(azimuths_) && max(azimuths_) <= azimuth_bins_[-1]`

Also, we must be sure that `azimuth_bins_` are sorted, then I would suggest to sort them once they are pulled from the input file. 
Sure, contact me if I am not explaining myself!